### ISSUE TITLE
How to optimize a complex phase?

### ISSUE THREAD

**Issue by ziofil** [tensorflow/tensorflow#36958]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): yes
- OS Platform and Distribution: MacOS 10.15.3
- TensorFlow installed from binary: 2.1.0

**Describe the current behavior**
A real loss function that uses complex numbers internally requires Variables to be complex.
But this leads to complex gradients, even though they should be real.

**Describe the expected behavior**
In TF1.x I used to circumvent this problem by defining my real variables as (e.g.):
```python
x = tf.complex(tf.Variable(1.0, dtype=tf.float32), tf.constant(0.0, dtype=tf.float32))
```
TF2.x cannot compute gradients if I use this trick.

**Standalone code to reproduce the issue** 
For example, I want to optimize the _real_ angle `x` in the complex phase `exp(1j*x)`:

```python
def loss(x):
    return tf.abs(tf.exp(1j*x)-1.0) # minimized by x = 0.0

x = tf.Variable(1.0, dtype=tf.complex64) # forced to have type tf.complex64
LR = 0.01

for n in range(10):
    with tf.GradientTape() as tape:
        L = loss(x)
        grad = tape.gradient(L, x)
        x.assign_sub(LR * grad)

print(x)
# <tf.Variable 'Variable:0' shape=() dtype=complex64, numpy=(0.9122518+0.043542963j)>
```

**Comment by gadagashwini** [comment]
I was able to reproduce the issue with Tf 2.1.
Please take a look at the [gist](https://colab.sandbox.google.com/gist/gadagashwini/8aff9ca30ab54aae040b1a2e3628f0bc/untitled402.ipynb). Thanks

**Comment by allenlavoie** [comment]
As long as you do it inside the `GradientTape`, the same trick should work fine in 2.x:

```
import tensorflow as tf
v = tf.Variable(1.0, dtype=tf.float32)
with tf.GradientTape() as tape:
  x = tf.complex(v, tf.constant(0.0, dtype=tf.float32)) ** 2.
v_grad = tape.gradient(x, v)
print(v_grad)  # tf.Tensor(2.0, shape=(), dtype=float32)
```

In 1.x it probably didn't matter where you put the `tf.complex`, but with `GradientTape` it really needs to be inside the scope so the tape can record the operation. But otherwise it'll be treated identically as with `tf.gradients`.

Does that work for you?

**Comment by ziofil** [comment]
It does! Thank you so much
